docs/design/language-support-notebook.md

```diff
@@ -0,0 +1,303 @@
+LSP 3.17 expects LSP servers to syncronize the notebook contents by these notifications, and provide language features based on the synchronized content.
+
+# Problems
+
+Though LSP 3.17 provides some specifications around notebook support, it doesn't work out of the box with Metals for mainly two reasons.
+
+## Metals doesn't understand the Scala program in a cell
+
+As you found, LSP client (VSCode) sends `textDocument/*` request with
+
+- `uri` points to the notebook cell.
+- `position` represents the position in the cell.
+
+```
+"textDocument": {
+  "uri": "vscode-notebook-cell:/Users/tanishiking/src/github.com/tanishiking/scala3-playground/src/main/scala/Untitled-1.ipynb#W2sZmlsZQ%3D%3D"
+},
+"position": {
+  "line": 0,
+  "character": 2
+}
+```
+
+However, Metals can't understand each cell of notebook. For example, in the following setting, Metals can't handle cell2 and cell3 (in Scala2, Metals can't cell1 too). Because each cell doesn't compile.
+
+```scala
+// cell1
+def add(a: Int, b: Int): Int = a + b
+
+// cell2
+add(1, 2)
+
+// cell3
+import $ivy.`io.circe::circe-generic:0.11.1`
+```
+
+**In order to analyze Scala code and provide compiler-based language features, Metals have to convert cells into compilable Scala program.**
+
+## Metals server cannot recognise `*.ipynb` file as a compile target
+
+Metals server provide language features to the files that are recognized as a `buildTarget/sources` of [Build Server Protocol](https://build-server-protocol.github.io/docs/specification.html#build-target-sources-request).
+
+However, current Metals doesn't recognize `*.ipynb` (or combined Scala source, that we'll mention later) for Build Servers.
+
+**Metals have to run build server that recognizes `*.ipynb` files as a `SourceItem`**. Otherwise, Metals cannot tell which Scala version of presentation compiler should run against the notebook (Kernel).
+
+# Solution / Technical Architecture
+
+In order to provide language support for notebooks, two things are needed.
+
+- Convert cells to compilable Scala sources.
+  - Also, Metals have to convert the incoming request parameter (for cells) into the parameter for the (converted) compilable Scala source (back and forth).
+  - Converted sources have to be on the filesystem, because Metals expect sources to be on the filesystem.
+- Build Server that recognizes `*.ipynb` as a build target.
+  - Where `ScalaBuildTarget.scalaVersion` should match with the Scala version of the running Kernel.
+  - When users open a notebook with Scala Kernel, Metals should prompt running the Build Server for the notebook, like Metals does for Ammonite script.
+
+## Components
+
+- `Almond BSP`
+  - Run on the same process with `Almond Kernel`, but the I/O will be via unix domain socket at a pre-defined location (somewhere in `~/.local/share/almond/bsp/{kernel-name}.sock`). See: https://github.com/scalameta/metals/pull/4253#issuecomment-1222091839
```

Comment:
Thinking a bit further about this (that I proposed in https://github.com/scalameta/metals/pull/4253#issuecomment-1222091839), I'm very slightly worried about how Metals will be able to get the PID of the Almond process corresponding to a given notebook, or more generally how Metals can find a way to connect to the right Almond kernel.

I skimmed through the [notebook section of LSP 3.17](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notebookCellTextDocumentFilter) (that you linked above), and couldn't find any way of letting servers know about the kernel processes that the LSP client starts.

To work around that, maybe we could have hashed the notebook path, and used that hash instead of the kernel PID. But the kernel [doesn't know about the path of the notebook it is asked to run](https://jupyter-client.readthedocs.io/en/stable/messaging.html), it only knows about individual code cells passed to it via [Jupyter message protocol](https://jupyter-client.readthedocs.io/en/stable/messaging.html) messages. So hashing the notebook path isn't an option either.

Another workaround could consist in having Almond leave its PID in the metadata of some [display data](https://jupyter-client.readthedocs.io/en/stable/messaging.html#display-data) or [execute_result](https://jupyter-client.readthedocs.io/en/stable/messaging.html#id6). That might be a bit flaky though, as for the PID to be written in the notebook file, this would require at least a cell to have run. Not to add too much overhead, we might want to return those metadata only for the first cell say, but then the metadata would go away if users delete that cell… (and the kernel doesn't know about cell deletion…) So it might need to be put in all execute_result metadata.

Also, those metadata, once written on disk, might be stale (they may correspond to the kernel that previously ran the notebook).

Don't know if asking people involved in LSP to update the protocol, so that LSP servers could know about Jupyter kernels started by LSP clients, could be an option here…

Reply:
>  I'm very slightly worried about how Metals will be able to get the PID of the Almond process corresponding to a given notebook,

Yeah, I think it's difficult (or even impossible) to find a Kernel PID from Metals. That's why I'm thinking maybe we can use `kernel name` or something instead of PID (more details later)

---

> To work around that, maybe we could have hashed the notebook path, and used that hash instead of the kernel PID.
> Another workaround could consist in having Almond leave its PID in the metadata of some [display data](https://jupyter-client.readthedocs.io/en/stable/messaging.html#display-data) or [execute_result](https://jupyter-client.readthedocs.io/en/stable/messaging.html#id6) (AFAIK, these are persisted in the notebook).

I think it works :+1:

> this would require at least a cell to have run.

Not really, we should be able to get kernel's metadata like on `notebookDocument/didOpen`, and we'll have access to that information without running cells

```json
 "metadata": {
  "kernelspec": {
   "display_name": "Scala",
   "language": "scala",
   "name": "scala3"
  },
  "language_info": {
   "codemirror_mode": "text/x-scala",
   "file_extension": ".sc",
   "mimetype": "text/x-scala",
   "name": "scala",
   "nbconvert_exporter": "script",
   "version": "2.13.8"
  }
 },
```

On the other hand, what do you think about using the Kernel spec name instead of embedding PID information to the name or display name? 

One concern is: is the kernel name unique?; can we say there won't be multiple same kernels running at the same time?

---

> Don't know if asking people involved in LSP to update the protocol, so that LSP servers could know about Jupyter kernels started by LSP clients, could be an option here…

I think LSP's responsibility is just to provide some interfaces between the editor(notebook) and language servers and how to connect with Kernel is out of the scope of LSP, I guess.

(kinda off-topic: the notebook LSP spec seems like the notebook content can be evaluated purely with its content (by naively concatenating cells into a script), and assuming there's no need to connect with Kernel).